src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -0,0 +1,253 @@
+    kWPAPersonal  = 2,
+    kWPA2Personal = 3,
+    kWPA3Personal = 4,
+};
+
+struct ThreadNetworkInfo
+{
+    uint8_t mDataset[kMaxThreadDatasetLen];
+    uint8_t mDatasetLen;
+};
+
+struct WiFiNetworkInfo
+{
+    uint8_t mSSID[kMaxWiFiSSIDLen + 1];
+    uint8_t mSSIDLen;
+    uint8_t mCredentials[kMaxWiFiCredentialsLen];
+    uint8_t mCredentialsLen;
+    WiFiSecurityType mSecurity;
+};
+
+struct NetworkInfo
+{
+    uint8_t mNetworkID[kMaxNetworkIDLen];
+    uint8_t mNetworkIDLen;
+    uint8_t mEnabled;
+    NetworkType mNetworkType;
+    union NetworkData
+    {
+        ThreadNetworkInfo mThread;
+        WiFiNetworkInfo mWiFi;
+    } mData;
+};
+
+namespace {
+NetworkInfo sNetworks[kMaxNetworks];
+}
+
+void OnAddThreadNetworkCommandCallback(app::Command *, EndpointId, const uint8_t * operationalDataset, uint64_t breadcrumb,
+                                       uint32_t timeoutMs)
+{
+    CHIP_ERROR err                = CHIP_ERROR_NO_MEMORY;
+    size_t operationalDataset_Len = strlen(reinterpret_cast<const char *>(operationalDataset));
+
+    for (size_t i = 0; i < kMaxNetworks; i++)
+    {
+        if (sNetworks[i].mNetworkType == NetworkType::kUndefined)
+        {
+            memcpy(sNetworks[i].mData.mThread.mDataset, operationalDataset, operationalDataset_Len);
+            sNetworks[i].mData.mThread.mDatasetLen = operationalDataset_Len;
+
+            sNetworks[i].mNetworkType = NetworkType::kThread;
+            sNetworks[i].mEnabled     = false;
+
+            memcpy(sNetworks[i].mNetworkID, kThreadNetworkName, sizeof(kThreadNetworkName));
+            sNetworks[i].mNetworkIDLen = sizeof(kThreadNetworkName);
+
+            err = CHIP_NO_ERROR;
+            break;
+        }
+    }
+
+    SuccessOrExit(err);
+exit:
+    ChipLogFunctError(err);
+}
+
+void OnAddWiFiNetworkCommandCallback(app::Command *, EndpointId, const uint8_t * ssid, const uint8_t * credentials,
+                                     uint64_t breadcrumb, uint32_t timeoutMs)
+{
+    CHIP_ERROR err         = CHIP_ERROR_NO_MEMORY;
+    size_t ssid_Len        = strlen(reinterpret_cast<const char *>(ssid));
+    size_t credentials_Len = strlen(reinterpret_cast<const char *>(credentials));
+
+    for (size_t i = 0; i < kMaxNetworks; i++)
+    {
+        if (sNetworks[i].mNetworkType == NetworkType::kUndefined)
+        {
+            memcpy(sNetworks[i].mData.mWiFi.mSSID, ssid, ssid_Len);
+            sNetworks[i].mData.mWiFi.mSSIDLen = ssid_Len;
+            memcpy(sNetworks[i].mData.mWiFi.mCredentials, credentials, credentials_Len);
+            sNetworks[i].mData.mWiFi.mCredentialsLen = credentials_Len;
+            memcpy(sNetworks[i].mNetworkID, sNetworks[i].mData.mWiFi.mSSID, ssid_Len);
+            sNetworks[i].mNetworkIDLen = ssid_Len;
+
+            sNetworks[i].mNetworkType = NetworkType::kWiFi;
+            sNetworks[i].mEnabled     = false;
+
+            err = CHIP_NO_ERROR;
+            break;
+        }
+    }
+
+    SuccessOrExit(err);
```

Comment:
nit: I believe the exit pattern in this case makes the code harder to read. This could be:

```cpp
for (size_t i = 0; i < kMaxNetworks; i++) 
{
    if (sNetworks[i].mNetworkType != NetworkType::kUndefined) {
        continue;
    }
    
    memcpy(sNetworks[i].mData.mWiFi.mSSID, ssid, ssid_Len);
￼    sNetworks[i].mData.mWiFi.mSSIDLen = ssid_Len;

    // ....
    
    ChipLogDetail(Zcl, "WiFi provisioning data: SSID: %s", ssid);
    return CHIP_NO_ERROR;
}
 
return CHIP_ERROR_NO_MEMORY;
```

Reply:
Updated, thanks.